chemprop/models/ffn.py

```diff
@@ -0,0 +1,439 @@
+        :param ffn_base: The shared base layers (all but the last) of the FFN between tasks.
+        :param ffn_type: The type of target (atom or bond).
+        :param shared_ffn: Whether to share weights in the ffn between different atom tasks and bond tasks.
+        """
+        super().__init__()
+
+        if num_layers == 1:
+            base_output_size = features_size
+        else:
+            base_output_size = hidden_size
+
+        if shared_ffn:
+            self.ffn_readout = nn.Sequential(
+                ffn_base,
+                build_ffn(
+                    first_linear_dim=base_output_size,
+                    hidden_size=hidden_size,
+                    num_layers=1,
+                    output_size=output_size,
+                    dropout=dropout,
+                    activation=activation,
+                ),
+            )
+        else:
+            self.ffn_readout = build_ffn(
+                first_linear_dim=features_size,
+                hidden_size=hidden_size,
+                num_layers=num_layers,
+                output_size=output_size,
+                dropout=dropout,
+                activation=activation,
+            )
+        self.ffn_type = ffn_type
+
+    def forward(
+        self,
+        input: Tuple[torch.Tensor, List, torch.Tensor, List, torch.Tensor],
+        bond_types: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Runs the :class:`FFN` on input.
+        :param input: A tuple of atom and bond informations of each molecule.
+        :param constraints: A PyTorch tensor which applies constraint on atomic/bond properties.
+        :param bond_types: A PyTorch tensor storing bond types of each bond determined by RDKit molecules.
+        :return: The output of the :class:`FFN`, a PyTorch tensor containing a list of property predictions.
+        """
+        a_hidden, a_scope, b_hidden, b_scope, b2br = input
+
+        if self.ffn_type == "atom":
+            hidden = a_hidden
+            scope = a_scope
+        elif self.ffn_type == "bond":
+            forward_bond = b_hidden[b2br[:, 0]]
+            backward_bond = b_hidden[b2br[:, 1]]
+            scope = [((start - 1) // 2, size // 2) for start, size in b_scope]
+        else:
+            raise RuntimeError(f"Unrecognized ffn_type of {self.ffn_type}.")
+
+        if self.ffn_type == "atom":
+            output = self.ffn_readout(hidden)
+            output = output[1:]  # remove the first one which is zero padding
+        elif self.ffn_type == "bond":
+            b_hidden_1 = torch.cat([forward_bond, backward_bond], dim=1)
+            b_hidden_2 = torch.cat([backward_bond, forward_bond], dim=1)
+            output_1 = self.ffn_readout(b_hidden_1)
+            output_2 = self.ffn_readout(b_hidden_2)
+            output = (output_1 + output_2) / 2
+            if bond_types is not None:
+                output = output + bond_types.reshape(-1, 1)
+
+        return output
+
+class FFNAtten(nn.Module):
```

Comment:
this class can be defined as a subclass of `FFN` to avoid code duplication

Reply:
I think those two classes are quite different either in __init__() and forward(). There is not too much code duplication in those two classes.

Reply:
It's only "different" because you've organized them be that way. `FFNAtten` is a logical extension of `FFN` if you reorganize it correctly:
```python
class FFN(nn.Module):
    def __init__(
        self,
        features_size: int,
        hidden_size: int,
        num_layers: int,
        output_size: int,
        dropout: float,
        activation: str,
        ffn_base: Optional[nn.Module] = None,
        ffn_type: str = "atom",
    ):
        super().__init__()

        base_output_size = features_size if num_layers == 1 else hidden_size

        if ffn_base:
            self.ffn = ffn_base
        else:
            if num_layers > 1:
                self.ffn = nn.Sequential(
                    build_ffn(
                        first_linear_dim=features_size,
                        hidden_size=hidden_size,
                        num_layers=num_layers - 1,
                        output_size=hidden_size,
                        dropout=dropout,
                        activation=activation,
                    ),
                    get_activation_function(activation),
                )
            else:
                self.ffn = nn.Identity()
        
        self.ffn_readout = nn.Sequential(
            nn.Dropout(dropout),
            nn.Linear(base_output_size, output_size),
        )
        self.ffn_type = ffn_type
```

which can now be seen to be the logical superclass of `FFNAtten`:
```python
class FFNAtten(nn.Module):
    def __init__(
        self,
        features_size: int,
        hidden_size: int,
        num_layers: int,
        output_size: int,
        dropout: float,
        activation: str,
        ffn_base: Optional[nn.Module] = None,
        ffn_type: str = "atom",
        weights_ffn_num_layers: int = 2,
    ):
        super().__init__()

        base_output_size = features_size if num_layers == 1 else hidden_size

        if ffn_base:
            self.ffn = ffn_base
        else:
            if num_layers > 1:
                self.ffn = nn.Sequential(
                    build_ffn(
                        first_linear_dim=features_size,
                        hidden_size=hidden_size,
                        num_layers=num_layers - 1,
                        output_size=hidden_size,
                        dropout=dropout,
                        activation=activation,
                    ),
                    get_activation_function(activation),
                )
            else:
                self.ffn = nn.Identity()
        # build a linear readout from the hidden outputs
        self.ffn_readout = nn.Sequential(
            nn.Dropout(dropout),
            nn.Linear(base_output_size, output_size),
        )
        self.ffn_type = ffn_type
        self.weights_readout = build_ffn(
            first_linear_dim=base_output_size,
            hidden_size=hidden_size,
            output_size=1,
            num_layers=weights_ffn_num_layers,
            dropout=dropout,
            activation=activation,
        )
```
and rewritten using inheritance:
```python
class FFNAtten(FFN):
    def __init__(
        self,
        features_size: int,
        hidden_size: int,
        num_layers: int,
        output_size: int,
        dropout: float,
        activation: str,
        ffn_base: Optional[nn.Module] = None,
        ffn_type: str = "atom",
        weights_ffn_num_layers: int = 2,
    ):
       super().__init__(
            features_size,
            hidden_size,
            num_layers,
            output_size,
            dropout,
            activation,
            ffn_base,
            ffn_type
        )
        self.weights_readout = build_ffn(
            first_linear_dim=base_output_size,
            hidden_size=hidden_size,
            output_size=1,
            num_layers=weights_ffn_num_layers,
            dropout=dropout,
            activation=activation,
        )
```
the same can be done for `forward()` (with a bit more work). The reason this matters is inheritance makes it easier to reason about logical relationships between classes/methods. You can see this even in the figure you've created to explain atom/bond-level target prediction. We use the _same_ base FFN for both constrained and unconstrained predictions, with some seasoning **on top** for constrained predictions. Therefore, we should write the code to reflect how we reason about the concept.

Reply:
you can also refactor `FFN.forward()` to be comprised of two calls: (1) `calc_hidden()`, which returns the initial hidden outputs and (2) `readout`, which takes the hidden outputs to a final output. `FFNAtten` would then only need to override `readout` because both networks calculate the initial hiddens in the same and only differ in how they take these hiddens to a final output. Again, this reflects how we _present_ the concept to readers in both text and figures.

Reply:
Thanks for guiding me through this. I made the change now.